### ISSUE TITLE
askama with markdown filter using String: cannot move out of `self.name` which is behind a shared reference

### ISSUE THREAD

**Issue by schaschko** [djc/askama#719]
I cloned the repo and then run this test:

https://github.com/djc/askama/blob/5748c357d435b24848d1571df010d777859fede9/askama_axum/tests/basic.rs

This works, then I modify basic.rs to (`String` instead of `&str`):
```
use askama::Template;
use axum::{
    body::Body,
    http::{Request, StatusCode},
    routing::get,
    Router,
};
use tower::util::ServiceExt;

#[derive(Template)]
#[template(path = "hello.html")]
struct HelloTemplate {
    name: String,
}

async fn hello() -> HelloTemplate {
    HelloTemplate { name: String::from("world") }
}

#[tokio::test]
async fn template_to_response() {
    let app = Router::new().route("/", get(hello));

    let res = app
        .oneshot(Request::builder().uri("/").body(Body::empty()).unwrap())
        .await
        .unwrap();
    assert_eq!(res.status(), StatusCode::OK);

    let headers = res.headers();
    assert_eq!(headers["Content-Type"], "text/html; charset=utf-8");

    let body = hyper::body::to_bytes(res.into_body()).await.unwrap();
    // assert_eq!(&body[..], b"Hello, world!");
}
```

Still works. Then I modify https://github.com/djc/askama/blob/5748c357d435b24848d1571df010d777859fede9/askama_axum/templates/hello.html to:
```
Hello, {{ name|markdown }}!
```

If I don't use the markdown filter, e.g. `Hello, {{ name }}!`, then it works. How to fix this error?

My goal was to be able to read in markdown files during runtime (with `fs::read_to_string( >path-to-markdown-file.md< )` which gives `String`).

Thank you

**Comment by djc** [comment]
I think this is because `askama::filters::markdown` takes `S: AsRef<str>` when it should probably really just be taking `&str`. Do you want to try if that improves things, and if so, submit a PR?

**Comment by plainlevel** [comment]
```
async fn my_posts(State(pool): State<PgPool>,) -> impl IntoResponse {
    #[derive(FromRow)]
    struct Post {
        title: String,
        post_body: String,
    }

    let my_post = sqlx::query_as::<_, Post>("select title, post_body from my_posts")
        .fetch_all(&pool)
        .await
        .unwrap();

    // my template looks like this:
    // #[derive(Template)]
    // #[template(path = "post.html"]
    // pub struct MyTemplate {
    //     pub post_title: String,
    //     pub post_body: String,
    // }

    let template = MyTemplate {post_title: my_post[0].title, post_body: my_post[0].post_body};

    HtmlTemplate(template)
}
```
I'm wresting with Rust's compiler to get this thing to work. If anyone can answer these questions, I'd be very much grateful. 

1. How do I correctly load my template into markdown at the start of the program so that I don't have to make an sql query every time a user visits one blog post page. My markdown (post_body: String) is stored in PostgreSQL as a TEXT type. 

2. I can't figure out how to convert String (which my sql query returns) into `&'a str` or `AsRef<str>`. I'm completely lost as to what markdown wants as a variable.

**Comment by plainlevel** [comment]
Edit, never mind I figured it out.

```
// I tried originally to get the values to be &'a str to make markdown happy, but
// I would only get the error: Value does not live long enough. So I tried String values.
#[derive(Clone)]
struct AppState {
    title: String,
    poist_body: String, 
} 

// in async main(), I query the database and create a state using the returned values
let shared_state = Arc::new(AppState{title: MYPOST[0].title.clone(), post_body: MYPOST[0].post_body.clone()});

// Now I simply call template.render() directly in the function and it works. Don't ask me how though. 
async fn my_posts(State(state): State<Arc<AppState>>) -> impl IntoResponse {
    let x = &state.title;
    let z = &state.post_body;
    let template = MyTemplate{post_title: &x, post_body: &z};
    match template.render() {
            Ok(html) => Html(html).into_response(),
            Err(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Failed to render template. Error {}", err),
            ).into_response(),
    }
}
```

**Comment by djc** [comment]
Happy that you figured it out!

**Comment by chase-lambert** [comment]
I'm trying to get past this issue. I tried to switch my struct to using an `&str` but that creates a lot of lifetime issues and according to this SO conversation, they recommend not doing so: https://stackoverflow.com/questions/76140189/higher-ranked-lifetime-error-for-axum-post

Above you mentioned that 

> askama::filters::markdown takes S: AsRef<str> when it should probably really just be taking &str 

How do I actually do this? Do I need to fork the repo and try to make that change? But then how do I use that forked code in my Rust app to test it out?

**Comment by chase-lambert** [comment]
So I forked it and made the following small changes:
```
 λ ~/projects/rust/askama/askama : git diff
diff --git a/askama/src/filters/mod.rs b/askama/src/filters/mod.rs
index a7aa62b..3f18655 100644
--- a/askama/src/filters/mod.rs
+++ b/askama/src/filters/mod.rs
@@ -349,7 +349,7 @@ pub fn markdown<E, S>(
 ) -> Result<MarkupDisplay<E, String>>
 where
     E: Escaper,
-    S: AsRef<str>,
+    S: Into<String>,
 {
     use comrak::{
         markdown_to_html, ComrakExtensionOptions, ComrakOptions, ComrakParseOptions,
@@ -389,7 +389,7 @@ where
         },
     };

-    let s = markdown_to_html(s.as_ref(), options.unwrap_or(&DEFAULT_OPTIONS));
+    let s = markdown_to_html(s.into().as_ref(), options.unwrap_or(&DEFAULT_OPTIONS));
     Ok(MarkupDisplay::new_safe(s, e))
 }
```

I think this might be doing what I want but I don't actually know how to test that in my own project to see if it does what I hope it does.  Do you think this would allow you to have a Struct field be a `String` and be able to call `{{ foo|markdown }}` on it (while also still allowing a `&str` if so desired)? 

All the Askama repo tests still pass but I didn't see anything `markdown` specific in the tests.

**Comment by djc** [comment]
It's not super-clear to me what problem you're trying to solve. Do you have a minimal reproduction?

**Comment by chase-lambert** [comment]
Yes, I can try and create a minimal repo. Sorry for the confusion.

I'm reaching out to an API and serializing the response with Serde and pushing that out to the template. Those structs that mirror the API json response use `Strings` as I find it much easier to work with instead of `&strs` for lifetime reasons. But when I change the template output from `{{ message.content }}` to `{{ message.content|markdown }}` I get the following error: 

```error[E0507]: cannot move out of `message.content` which is behind a shared reference
  --> src/handlers/lesson.rs:21:28
   |
21 | #[derive(serde::Serialize, Template)]
   |                            ^^^^^^^^ move occurs because `message.content` has type `std::string::String`, which does not implement the `Copy` trait
   |
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)```

I'm confused on the proper solution to fix this and I thought it would be by changing all my structs to `&str` from `String` but that is going to take a lot of effort.

My fix so far is to use the comrak library myself: `let ai_response_html = markdown_to_html(&ai_response, &ComrakOptions::default());` and then in the template call it with `{{ message.content|safe }}` but was curious on how to just use the intended Askama markdown filter instead.
